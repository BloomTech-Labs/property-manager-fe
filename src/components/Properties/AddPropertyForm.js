/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable react/prop-types */
/** @jsx jsx */
import { Formik, Field, Form } from 'formik';
import * as Yup from 'yup';
import { jsx } from '@emotion/core';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import { makeStyles } from '@material-ui/core/styles';
import FormErrors from '../../helpers/FormErrors';

const useStyles = makeStyles(theme => ({
  formControl: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
    width: '100%',
    maxWidth: '600px'
  }
}));

const validationSchema = Yup.object().shape({
  name: Yup.string()
    .max(255, 'Name entered was too long')
    .required('Must enter a Property Name'),

  street: Yup.string()
    .max(255, 'Address entered was too long')
    .required('Must enter a street address'),
  city: Yup.string()
    .max(50, 'City entered was too long')
    .required('Must enter a city'),
  zip: Yup.string()
    .min(5, 'Must enter 5-digit zip code')
    .max(5, 'Must enter 5-digit zip code')
    .required('Must enter a 5-digit zip code'),
  state: Yup.string()
    .max(50, 'State entered was too long')
    .required('Must enter the state'),
  status: Yup.string().required('Property Status is required!')
});

export default function AddPropertyForm({ submit }) {
  const classes = useStyles();

  return (
    <div className="form-card">
      <h2>Add Property</h2>
      <Formik
        validationSchema={validationSchema}
        initialValues={{
          name: '',
          street: '',
          city: '',
          state: '',
          zip: '',
          status: ''
        }}
        onSubmit={values => {
          // console.log(values);
          submit(values);
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form data-testid="form-element">
            <div className="input-wrapper">
              <label htmlFor="name">Property Name</label>
              <Field
                placeholder="Enter a name for your Property"
                name="name"
                type="text"
              />
              <FormErrors touched={touched.name} message={errors.name} />
            </div>

            <div className="input-wrapper">
              <label htmlFor="street">Street Address</label>
              <Field placeholder="Street address" name="street" type="text" />
              <FormErrors touched={touched.street} message={errors.street} />
            </div>

            <div className="input-wrapper">
              <label htmlFor="city">City</label>
              <Field placeholder="City" name="city" type="text" />
              <FormErrors touched={touched.city} message={errors.city} />
            </div>

            <div className="input-wrapper">
              <label htmlFor="zip">Zip Code</label>
              <Field
                placeholder="Enter a 5-digit Zip Code"
                name="zip"
                type="number"
              />

              <FormErrors touched={touched.zip} message={errors.zip} />
            </div>

            <div className="input-wrapper">
              <label htmlFor="state">State</label>
              <Field placeholder="State" name="state" type="text" />
              <FormErrors touched={touched.state} message={errors.state} />
            </div>

            <FormControl className={classes.formControl}>
              <InputLabel>Property Status</InputLabel>
              <Field name="status" as={Select}>
                <MenuItem value="vacant">Vacant</MenuItem>
                <MenuItem value="occupied">Occupied</MenuItem>
              </Field>
              <FormErrors touched={touched.status} message={errors.status} />
            </FormControl>

            <div className="submit-btn-wrapper">
              <button
                className="btn btn-animated"
                type="submit"
                disabled={isSubmitting}
              >
                Submit
              </button>
            </div>
          </Form>
        )}
      </Formik>
    </div>
  );
}
