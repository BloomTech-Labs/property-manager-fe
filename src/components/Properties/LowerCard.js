import React from 'react';
import { navigate } from '@reach/router';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import { Skeleton } from '@material-ui/lab';

function LowerCard({ props }) {
  const { icon, iconPath } = props;

  return (
    <>
      <CardActionArea
        onClick={() => {
          navigate(iconPath);
        }}
      >
        <CardActions className="property-card-action">
          {icon || <Skeleton variant="circle" width={28} height={28} />}
        </CardActions>
      </CardActionArea>
    </>
  );
}

export default LowerCard;
