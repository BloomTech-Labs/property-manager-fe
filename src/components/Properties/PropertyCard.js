/* eslint react/prop-types: 0 */
import React from 'react';
// Router
import { navigate } from '@reach/router';
// MUI Imports
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Divider from '@material-ui/core/Divider';

// Defaults
import { Skeleton } from '@material-ui/lab';

// Component Styling
const useStyles = makeStyles({
  card: {
    maxWidth: 345
  },
  media: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    '& svg': {
      height: 140
    }
  },
  cardAction: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    '& svg': {
      fontSize: '2rem',
      color: '#76e276',
      transition: '0.3s'
    },
    '&:hover': {
      '& > svg': {
        color: '#2d3b4f'
      }
    }
  }
});

// This component represents a single PropertyCard
//
// If the card is used with no props it will
// default to placeholder.
//
// Properties:
//
// props.title: string | default: <Skeleton />
// props.icon: component | default: <Skeleton />
// props.svg: component | default: <Skeleton />
//
// WIP: Make the CardMedia dynamic to accept an image
// instead of only an SVG component as a property.

export default function PropertyCard(props) {
  const classes = useStyles();

  const { icon, title, svg, handleOpen, property, upperPath, iconPath } = props;

  // event handler for clicking on the upper CardActionArea
  const handleClick = () => {
    // conditional logic to check if handleOpen property exists
    if (handleOpen) {
      // We call the handleOpen() function passed down as a prop
      // and give it the property that was passed down as a prop.
      //
      // This gives us access to the property object data that we
      // can use in the individual property Modal
      return handleOpen(property);
    }

    return null;
  };

  return (
    <Grid item xs={12} sm={6} md={4} lg={3}>
      <Card className={classes.card}>
        <CardActionArea
          onClick={() => {
            navigate(upperPath);

            handleClick();
          }}
        >
          <CardMedia className={classes.media}>
            {svg || <Skeleton variant="rect" height={140} width="100%" />}
          </CardMedia>
          <CardContent>
            <Typography gutterBottom variant="h5" component="h2">
              {title || <Skeleton />}
            </Typography>
          </CardContent>
        </CardActionArea>
        <Divider />
        <CardActionArea
          onClick={() => {
            navigate(iconPath);
          }}
        >
          <CardActions className={classes.cardAction}>
            {icon || <Skeleton variant="circle" width={28} height={28} />}
          </CardActions>
        </CardActionArea>
      </Card>
    </Grid>
  );
}
