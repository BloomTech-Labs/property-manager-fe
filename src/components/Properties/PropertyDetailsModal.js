import React from 'react';
import PropTypes from 'prop-types';

// MUI
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import IconButton from '@material-ui/core/IconButton';
import Avatar from '@material-ui/core/Avatar';
import Divider from '@material-ui/core/Divider';
import CardMedia from '@material-ui/core/CardMedia';

// Icons
import { FaPen, FaHome } from 'react-icons/fa';
import { makeStyles } from '@material-ui/core/styles';

// Components
import { navigate } from '@reach/router';
import { Button } from '@material-ui/core';
import MuiModal from '../UI/MuiModal';
import LocationSVG from '../SVG/LocationSVG';

// Define styling for modal
const useStyles = makeStyles(theme => ({
  card: {
    width: 400
  },
  title: {
    margin: 0,
    lineHeight: 1
  },
  avatar: {
    backgroundColor: '#2d3b4f'
  },
  media: {
    padding: theme.spacing(2)
  },
  address: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    '& h4': {
      marginRight: theme.spacing(2)
    }
  },
  tenantInfo: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    '& button': {
      marginTop: theme.spacing(2)
    }
  }
}));

export default function PropertyDetailsModal({ property, open, close }) {
  // bring in access to custom styling
  const classes = useStyles();

  // Pull out data from the property object passed
  // in from Properties.js component
  const { id, city, name, state, status, street, zip } = property;
  // TODO: add in functionality to handle the image
  // from the property object and allow SVG default

  return (
    <MuiModal open={open} close={close}>
      <Card className={classes.card}>
        <CardHeader
          avatar={
            <Avatar className={classes.avatar}>
              <FaHome />
            </Avatar>
          }
          title={<h4 className={classes.title}>{name}</h4>}
          action={
            <IconButton
              onClick={() => navigate(`properties/edit/${id}`)}
              aria-label="edit"
            >
              <FaPen />
            </IconButton>
          }
        />
        <Divider />
        <CardMedia className={classes.media}>
          <LocationSVG />
        </CardMedia>
        <Divider />
        <CardContent className={classes.address}>
          <h4>Address:</h4>
          <div>
            <Typography variant="subtitle1">{street}</Typography>
            <Typography variant="body1">
              {city}, {state}, {zip}
            </Typography>
          </div>
        </CardContent>
        <Divider />
        <CardContent className={classes.tenantInfo}>
          <h6 style={{ textAlign: 'center' }}>Status: {status}</h6>
          <Button
            color="primary"
            variant="contained"
            onClick={() => navigate('tenant/add')}
          >
            Add Tenant
          </Button>
        </CardContent>
      </Card>
    </MuiModal>
  );
}

PropertyDetailsModal.propTypes = {
  property: PropTypes.oneOfType([
    PropTypes.shape({
      city: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      state: PropTypes.string.isRequired,
      status: PropTypes.string.isRequired,
      street: PropTypes.string.isRequired,
      zip: PropTypes.string.isRequired,
      image: PropTypes.oneOfType([
        PropTypes.string.isRequired,
        PropTypes.instanceOf(null).isRequired
      ])
    }),
    PropTypes.object.isRequired
  ]).isRequired,
  open: PropTypes.bool.isRequired,
  close: PropTypes.func.isRequired
};
