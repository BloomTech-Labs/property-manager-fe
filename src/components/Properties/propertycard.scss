.property-card {
  max-width: 345px;
}
  
.property-card-media {
  display: flex;
  justify-content: center;
  align-items: center;

  img {
    height: 140px;
  }
}
  
.card-property-action {
  display: flex;
  justify-content: center;
  align-items: center;

  &:svg {
    font-size: 2rem;
    color: #76e276;
    transition: 0.3s
  }

  &:hover {
    &:svg {
      color: #2d3b4f;
    }
  }
}
