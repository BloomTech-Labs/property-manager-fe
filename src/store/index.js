import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import { getFirebase } from 'react-redux-firebase';
import reducer from './reducers';

const middleware =
  process.ENV === 'production'
    ? [thunk.withExtraArgument(getFirebase)]
    : [thunk.withExtraArgument(getFirebase), logger];

export const makeStore = state =>
  createStore(
    reducer,
    state,
    compose(
      applyMiddleware(...middleware),

      /* eslint-disable */
      window.__REDUX_DEVTOOLS_EXTENSION__
        ? window.__REDUX_DEVTOOLS_EXTENSION__()
        : f => f
    )
  );

export default makeStore();
