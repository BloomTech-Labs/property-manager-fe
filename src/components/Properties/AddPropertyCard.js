/* eslint react/prop-types: 0 */
import React from 'react';
import { FaPlus, FaExclamationCircle } from 'react-icons/fa';
import PropertyCard from './PropertyCard';
import AddHouseSVG from '../SVG/AddHouseSVG';
import ErrorSVG from '../SVG/ErrorSVG';

export default function AddPropertyCard(props) {
  const { propertyNum, isLoading, error } = props;

  if (error && !isLoading) {
    return (
      <PropertyCard
        icon={<FaExclamationCircle />}
        title="Uh oh! There was an error."
        svg={<ErrorSVG />}
      />
    );
  }

  if (propertyNum === 0 && !isLoading) {
    return (
      <PropertyCard
        icon={<FaPlus />}
        title="No properties, add one?"
        svg={<AddHouseSVG />}
      />
    );
  }
  if (propertyNum === 0 && isLoading) {
    return null;
  }
  return (
    <PropertyCard
      icon={<FaPlus />}
      title="Add a New Property?"
      svg={<AddHouseSVG />}
    />
  );
}
