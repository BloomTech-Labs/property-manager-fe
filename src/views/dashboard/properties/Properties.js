/* eslint-disable no-lone-blocks */
import React, { useEffect, useState } from 'react';
// UI
import { Grid, Divider } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
// action
import { FaPen } from 'react-icons/fa';
import { getProperties } from '../../../store/actions';
import PropertyCard from '../../../components/Properties/PropertyCard';
import AddPropertyCard from '../../../components/Properties/AddPropertyCard';
import LocationSVG from '../../../components/SVG/LocationSVG';

export default function PropertyList() {
  const dispatch = useDispatch();

  const [loading, setLoading] = useState(true);

  const errMsg = useSelector(state => state.propReducer.errMsg);

  const propertyList = useSelector(state => state.propReducer.properties);

  useEffect(() => {
    setTimeout(() => {
      dispatch(
        getProperties(
          'https://pt6-propman-staging.herokuapp.com/api/properties'
        )
      ).then(() => {
        setLoading(false);
      });
    }, 2000);
  }, [dispatch]);

  return (
    <div className="properties">
      <h1>List of Properties</h1>
      <Divider />
      <br />

      <Grid container spacing={3}>
        {propertyList.map(property => {
          const { id, street } = property;

          return (
            <React.Fragment key={id}>
              <PropertyCard
                key={id}
                svg={<LocationSVG />}
                title={street}
                icon={<FaPen />}
              />
            </React.Fragment>
          );
        })}

        {(() => {
          if (propertyList.length > 0) {
            return null;
          }
          if (propertyList.length === 0 && loading === false) {
            return null;
          }
          return (
            <>
              <PropertyCard />
              <PropertyCard />
              <PropertyCard />
              <PropertyCard />
            </>
          );
        })()}

        <AddPropertyCard
          propertyNum={propertyList.length}
          isLoading={loading}
          error={errMsg}
        />
      </Grid>
    </div>
  );
}
