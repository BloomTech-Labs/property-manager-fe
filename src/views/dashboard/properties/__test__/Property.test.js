import React from 'react';
import * as rtl from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import store from '../../../../store';
import Properties from '../Properties';

afterEach(rtl.cleanup);

test('Properties renders', () => {
  const wrapper = rtl.render(
    <Provider store={store}>
      <Properties />
    </Provider>
  );

  expect(wrapper).toMatchSnapshot();
});
