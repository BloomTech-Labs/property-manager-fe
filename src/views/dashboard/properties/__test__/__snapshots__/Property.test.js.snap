// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properties renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="properties"
      >
        <h1>
          Properties
        </h1>
        <hr
          class="MuiDivider-root"
        />
        <br />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-7"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Add New
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <table
          aria-label="Property Table"
          class="MuiTable-root makeStyles-table-49"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                scope="col"
              >
                Id
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                Street
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                State
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                No. Tenants
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                scope="col"
              >
                Edit
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="properties"
    >
      <h1>
        Properties
      </h1>
      <hr
        class="MuiDivider-root"
      />
      <br />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-7"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          Add New
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <table
        aria-label="Property Table"
        class="MuiTable-root makeStyles-table-49"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            >
              Id
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              Street
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              State
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              No. Tenants
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
              scope="col"
            >
              Edit
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
